src/icp.rs

```diff
@@ -0,0 +1,168 @@
+use async_trait::async_trait;
+use ic_cdk::export::candid::{CandidType, Deserialize, Encode};
+use ic_cdk::export::Principal;
+use ic_ledger_types::{
+	query_archived_blocks, query_blocks, AccountIdentifier, Block, GetBlocksArgs, Operation,
+	Transaction, DEFAULT_SUBACCOUNT,
+};
+use std::collections::{BTreeMap, BTreeSet};
+
+pub const MAINNET_ICP_LEDGER: &str = "ryjl3-tyaaa-aaaaa-aaaba-cai";
+
+pub type Memo = u64;
+pub type BlockHeight = u64;
+
+pub struct Receiver<Q: TXQuerier> {
+	tx_querier: Q,
+	my_account: AccountIdentifier,
+	known_txs: BTreeSet<BlockHeight>, // set of block heights
+	unspent: BTreeMap<Memo, Amount>,  // received tokens per memo
+}
+
+#[async_trait]
+pub trait TXQuerier {
+	async fn query_tx(&self, block_height: u64) -> Option<TransactionNotification>;
+}
+
+#[derive(Default)]
+pub struct MockTXQuerier {
+	txs: BTreeMap<u64, TransactionNotification>,
+}
+
+#[async_trait]
+impl TXQuerier for MockTXQuerier {
+	async fn query_tx(&self, block_height: u64) -> Option<TransactionNotification> {
+		self.txs.get(&block_height).cloned()
+	}
+}
+
+pub struct CanisterTXQuerier {
+	icp_ledger: Principal,
+}
+
+#[async_trait]
+impl TXQuerier for CanisterTXQuerier {
+	async fn query_tx(&self, block_height: BlockHeight) -> Option<TransactionNotification> {
+		if let Some(block) = self.get_block_from_ledger(block_height).await {
+			return TransactionNotification::from_tx(block.transaction);
+		}
+		None
+	}
+}
+
+impl CanisterTXQuerier {
+	pub fn new(ledger: Principal) -> Self {
+		Self { icp_ledger: ledger }
+	}
+
+	async fn get_block_from_ledger(&self, block_height: BlockHeight) -> Option<Block> {
+		let args = GetBlocksArgs {
+			start: block_height,
+			length: 1,
+		};
+		if let Ok(result) = query_blocks(self.icp_ledger, args.clone()).await {
+			if result.blocks.len() != 0 {
+				return result.blocks.first().cloned();
+			}
+			if let Some(b) = result
+				.archived_blocks
+				.into_iter()
+				.find(|b| (b.start <= block_height && (block_height - b.start) < b.length))
+			{
+				if let Ok(Ok(range)) = query_archived_blocks(&b.callback, args).await {
+					return range.blocks.get((block_height - b.start) as usize).cloned();
+				}
+			}
+		}
+		None
+	}
+}
+
+impl<Q> Receiver<Q>
+where
+	Q: TXQuerier,
+{
+	pub fn new(q: Q, my_principal: Principal) -> Self {
+		Self {
+			tx_querier: q,
+			my_account: AccountIdentifier::new(&my_principal, &DEFAULT_SUBACCOUNT),
+			known_txs: Default::default(),
+			unspent: Default::default(),
+		}
+	}
+
+	/// Verifies a transaction, and if it is valid and new, tracks its funds.
+	pub async fn verify(&mut self, block_height: BlockHeight) -> bool {
+		if self.known_txs.contains(&block_height) {
+			return false;
+		}
+
+		if let Some(tx) = self.tx_querier.query_tx(block_height).await {
+			if !self.known_txs.insert(block_height) {
+				return false;
+			}
+			if tx.to != self.my_account {
+				return false;
+			}
+			*self.unspent.entry(tx.memo).or_insert(0.into()) += tx.get_amount();
```

Comment:
Where do you verify that the `tx.to` account is consistent with `tx.memo`?

Reply:
`tx.to` is the perun canister's account, and `tx.memo` is the first 8 bytes of the hash of the channel funding ID. They are completely unrelated.

Reply:
OK, and who sets `tx.memo` initially? Is this done off-chain when forming the tx?